### PR TITLE
Fix room unsubscribe

### PR DESCRIPTION
**Pull request by tnoho**
The situation when a client disconnect from the room when many clients connected in same room on the server.
At that situation many clients left in the room, but the server unsubscribe from room channel.

This patch confirm nobody left in the room before unsubscribe.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -217,15 +230,19 @@ function adapter(uri, opts){
 
     debug('removing %s from rooms %s', id, Object.keys(rooms));
 
-    Adapter.prototype.delAll.call(this, id);
```
Comment (fracmak) [inline index.js:220]: the delAll command was handling the code you've written below. 

See:

https://github.com/socketio/socket.io-adapter/blob/master/index.js#L80-L93

File: index.js [hunk 2]
```diff
@@ -200,8 +203,18 @@ function adapter(uri, opts){
 
   Redis.prototype.del = function(id, room, fn){
     debug('removing %s from %s', id, room);
+    var hasRoom = this.rooms.hasOwnProperty(room);
     Adapter.prototype.del.call(this, id, room);
-    runCommand('unsubscribe', this.getChannelName(room));
+    if (hasRoom && !this.rooms[room]) {
```
Comment (fracmak) [inline index.js:208]: Isn't this duplicative of Adapter's del() function?

https://github.com/socketio/socket.io-adapter/blob/master/index.js#L61-L70

Comment (tnoho) [inline index.js:208]: Yes, like that.
However, we have to check room is empty before unsubscribe room channel.
And, We have to call Adapter's del() function for manage Adapter's resource.
This implementation is as same as socket.io-redis.
https://github.com/socketio/socket.io-redis/blob/master/index.js#L175-L180

File: index.js (hunk 1 above)
Comment (tnoho) [inline index.js:220]: However we called del() function repeatedly rooms.
https://github.com/GannettDigital/socket.io-ioredis/pull/2/files#diff-168726dbe96b3ce427e7fedce31bb0bcR240
And del sids resource.
https://github.com/GannettDigital/socket.io-ioredis/pull/2/files#diff-168726dbe96b3ce427e7fedce31bb0bcR242
Those sequence is totally as same as Adapter's delAll() function.
So we don't have to call Adapter's delAll() function.
This implementation like socket.io-redis.
https://github.com/socketio/socket.io-redis/blob/master/index.js#L212-L220

File: index.js [hunk 3]
```diff
@@ -217,15 +230,19 @@ function adapter(uri, opts){
...
+        process.nextTick(fn.bind(null, null));
+      }
+      return;
     } else {
-      var self = this;
-      runCommand('unsubscribe', Object.keys(rooms).map(function(room) {
-        return self.getChannelName(room);
-      }));
+      for (var room in rooms) {
+        this.del(id, room);
```
Comment (tnoho) [inline index.js]: Oops, I found mistake.
var room is index of rooms.
We have to change this line to below.
```javascript
this.del(id, rooms[room]);
```

File: index.js (hunk 1 above)
Comment (fracmak) [inline index.js:220]: I disagree with socket.io-redis's implementation. Most of the logic they have is already in the socket-adapter base class and not necessary. Could you explain to me the situation you're having issues with?

Comment (tnoho) [inline index.js:220]: Your implementation of delAll() function don't check that any other sockets leave in the room.
https://github.com/GannettDigital/socket.io-ioredis/blob/master/index.js#L226
The situation of issue is written in top of this conversation. 
Adapter's delAll() function is remove only socketid from Room instance.
https://github.com/socketio/socket.io-adapter/blob/master/index.js#L85
Before unsubscribe I have to check that no sockets are left in the room.
This logic is as same as del() function.
https://github.com/GannettDigital/socket.io-ioredis/pull/2/files#diff-168726dbe96b3ce427e7fedce31bb0bcR240
https://github.com/GannettDigital/socket.io-ioredis/pull/2/files#diff-168726dbe96b3ce427e7fedce31bb0bcR204

My English is poor, but I appreciate your kindness.

Comment (fracmak) [inline index.js:220]: ah, I think I see what you're saying. Let me dig into this some more

**Comment by fracmak** [comment]
closing this in favor of #7

**Pull request closed without merging**